src/client/Client.cc

```diff
@@ -5416,9 +5416,13 @@ void Client::_try_to_trim_inode(Inode *in, bool sched_inval)
       if( in->ll_ref > 0 && sched_inval) {
         // FIXME: we play lots of unlink/link tricks when handling MDS replies,
         //        so in->dentries doesn't always reflect the state of kernel's dcache.
-        _schedule_invalidate_dentry_callback(dn, true);
+	if (!in->nlink)
+          _schedule_invalidate_dentry_callback(dn, true);
+	else
+          _schedule_invalidate_dentry_callback(dn, false);
       }
-      unlink(dn, true, true);
+      if (!in->nlink)
+        unlink(dn, true, true);
```

Comment:
These changes do not look related to `client: try to invalidate the dcache when any shared caps lost`. Can you explain? Perhaps in another commit?

Also, do you have a reproducer for the problem or is this something you believe exists but can't reproduce?

Reply:
> These changes do not look related to `client: try to invalidate the dcache when any shared caps lost`. Can you explain? Perhaps in another commit?

I intend to invalidate the dentry cache only, in such as kernel fuse or third part app, like the Ganesha's NFS, when any shared caps gets lost. To make sure that the kernel fuse or Ganesha's NFS to ignore the relevant dentry caches. 

There has two cases will trigger this in `handle_cap_grant()`, the first one is when `nlink == 0` and another is when any shared caps gets lost.

> 
> Also, do you have a reproducer for the problem or is this something you believe exists but can't reproduce?

I didn't have the Ganesha NFS setups, so tried to use the ceph-fuse by hard coding to enable the `entry_timeout` options for the lookup today. The test case is :

```
mounter-cephfs-A:   'touch /mnt/cephfs-A/file0; ll /mnt/cephfs-A/'
mounter-cpehfs-B:   'ln /mnt/cephfs-B/file0 /mnt/cephfs-B/file1; rm /mnt/cephfs-B/file0 -f'
mounter-cephfs-A:   'll /cephfs/file0'
```
Without this PR, in the last step in `mounter-cephfs-A` we can still see the `file0`.

More detail please see: https://tracker.ceph.com/issues/53730#note-5

IMO, this should be enough to help `Sheng Xie` to push the above tracker to add the `entry_timeout` and `attr_timeout` support, which really could improve the perf much.